### PR TITLE
OAuth upgrade path dialog and handling

### LINKED ISSUE #496: Add migration path for new FxA and Sync
Depends on #242

Need to let it fully bake. Have messaging for the user.

**Content:** 
Dialog Title: Reauthentication Required
Dialog Subtext: Due to a recent app update, we will need you to sign in again. Apologies for the inconvenience.
Dialog Action/Button: Continue

### Visual Design
https://mozilla.invisionapp.com/share/BQEUSO0S2#/311854834_Entry_List_-_OAuth_Messaging

### PR DESCRIPTION
**Pull request by sashei**
resolves #496 

#### expected behavior:
- When upgrading from an app that has the old login path, users will be sent to the welcome screen
- On the welcome screen, the login disclaimer dialog box will be shown.
- Hitting continue brings users to FxA signin

#### TESTING NOTES:

To test this branch, you need to have a clean install of pre-OAuth Lockbox. To make sure you have a clean install:

1. Disconnect your account inside the OAuth-signed-in app to make sure that the Keychain is empty when installing.
2. Delete the app, and then install an older version without OAuth (current production is #1664 so this will be the most accurate).
3. Sign in to this older version, confirm email etc so that you see logins syncing.
4. optionally lock or leave the app unlocked
5. Go to BB and install the most recent build of this branch
6. hopefully see the expected behavior?

### CODE REVIEW

**Comment by nickbrandt** [comment]
@sashei what is the best way to test this PR? I tried downloading the branch off BB and wasn't getting a dialog upon opening the app. Do I need to install an old version of the app, then download this branch to make it an update vs. a fresh install of the branch/app?

**Comment by sashei** [comment]
@nickbrandt that's exactly right; you need to install an older version of the app, login in the old way, and then update over that install with the one from this branch. This should only work if you have logged in the old way.

**Comment by devinreams** [comment]
@sashei can you help describe the expectations in a bit more detail? What's the "happy path" to hitting the upgrade? What scenarios should we try against?

I had an old version, upgraded over it, and when I re-opened the app I was on my entries screen with no prompt.

Then I manually locked and was thrown the prompt and had to hit continue.

Then it took me to a webview showing the "Firefox Accounts" management page (did not actually prompt me to sign in).

Some questions just from my quick experience:

- should this / is this prompt expected to fire on any app load on any screen?
- do we need any handling for clearing the auth state since I was somehow already signed in?
- if I don't actually authenticate when it takes me to the webview will the app just keep working? and never prompt me again?

**Comment by sashei** [comment]
Ok, I think the app is not forcing login on first launch after the upgrade. I will investigate that. In the meantime, the expected behavior is as follows:

- When upgrading from an app that has the old login path, users will be sent to the welcome screen.
- On the welcome screen, the login disclaimer dialog box will be shown.
- Hitting continue brings users to FxA signin (with email pre-populated)

Users opening the app for the first time will see the Welcome screen without a dialog box. Users that have already authenticated with the new OAuth flow will not get asked to reauthenticate.

**Comment by sashei** [comment]
One possible gotcha with testing: lockbox does not delete the new Keychain values when launching for the first time into the **old** login flow version. To ensure a clean device state for testing, make sure you have a **post-OAuth** install, and disconnect the linked FxA account via the account settings page. THEN delete the app and install a build with the old login flow to begin the testing steps as I detailed above for @nickbrandt.

**Comment by sashei** [comment]
ok, I think I fixed the routing behavior that y'all were seeing as well. good to re-test @nickbrandt @devinreams !!

**Comment by changecourse** [comment]
So I almost got the expected experience... Following Nick's lead, I did the following: 

-     Installed build w/ OAuth (1771)
-     Signed In > Then Disconnected Account > Deleted App
-     Installed build w/o OAuth (1557)
-     Signed in > Left Lockbox
-     Installed this branch
-     Open App > **Sent to Welcome Screen with Dialog** > **Hit Continue** 
-     **Autodirected to FxA Account Preferences, Appear to be Logged In**

**Comment by sashei** [comment]
@nickbrandt what version of iOS / phone are you using?

**Comment by linuxwolf** [comment]
This is pure speculation, but I wonder how much WKWebsiteDataStore is retained despite the app being deleted.

**Comment by sashei** [comment]
<i>/me increasingly annoyed that I didn't bring a test device home</i>

I think that has something to do with it @linuxwolf. I haven't been able to reproduce @nickbrandt 's results, but did find a cool *different* bug in which the app will still lock after 5 minutes and then let the user go to the entry list. Working through that one also.

**Comment by nickbrandt** [comment]
@sashei iPhone 7 running 11.4.1

**Comment by nickbrandt** [comment]
Tried testing again this morning, had the same results (no dialog) with the exception the 'Get Started' button now took me to the email first FxA screen as expected.

**Comment by changecourse** [comment]
Retesting this morning...

- Installed build w/ OAuth (1791)
- Signed In > Then Disconnected Account > Deleted App
- Installed build w/o OAuth (1540)
- Signed in > Taken to Confirmation Screen > Entries List >**Did Not Have to confirm email** > Left Lockbox
- Installed this branch
- Open App > **Routed to Welcome Screen with Dialog** > **Hit Continue** 
- **Now I'm autodirected to Entries List with "No Matching Entries" state**
- **Settings > My Account (No Avatar or email address displayed)** 
- **Tap "Disconnect Account" > Returns me to Entries List with "No Matching Entries" state**
- **Force Quit App > Reopen > Finally able to get back to Welcome Screen**
- Reauthenticating with FxA gets me back to a normal state

**Comment by devinreams** [comment]
I had luck with and saw the expected result with these steps:

1. download the new latest build from here (1776) as a Branch build and sign in, then Disconnect
2. download the old 1.1.1.1717 tag (Build 1718) as Branch build and sign in
3. then re-download the new latest build here (1776) which effectively upgrades over top
4. open the app, it showed entries list and sync spinner then _immediately_ drops me to the "Welcome" screen with the re-auth prompt ✅ 

img_0015

**Comment by sashei** [comment]
OK, this is still not working, but I've narrowed one issue down to some interesting `DataStore` behavior... when `reset`ing the DataStore, the `reset()` function never gets all the way through. It looks like `disconnect` does not succeed, and therefore the logins do not get removed and the `syncState` and `storageState` are never updated. We need the `syncState` and `storageState` values to be up to date to determine what state we're in as the app launches / goes through the routing. Maybe @jhugman can shed some light here...?

**Comment by sashei** [comment]
I suspect this might have something to do with the `UIApplication.shared` being different when we have upgraded, so the call to `self.fxaLoginHelper.applicationDidDisconnect(UIApplication.shared)` fails when we go to reset.

**Comment by sashei** [comment]
Ditto with this branch (implementation there, but tests are still broken for various reasons -- pushing a build out anyway). I would like @jhugman to take a look at the `DataStore.swift` changes that I made, but in the meantime I would appreciate more eyes on the functionality.

I'm doing some exploratory testing myself this afternoon and will post the results here. I really appreciate @nickbrandt @changecourse @devinreams 's patience working with this story and giving me such detailed reports!! I know it's super tedious to test :(

**Comment by sashei** [comment]
Testing notes:

Each time, started with the OAuth song and dance:
- Installed build w/ OAuth (1771)
- Signed In > Then Disconnected Account > Deleted App
- Installed build w/o OAuth (1557)
- Signed in

Follow-up tests:
--------------------------------
- Manually LOCKED application
- Installed this build
- Opened app > Upgrade dialog
- Selecting `Continue` brings user to signin
- Signed in successfully with OAuth ✅

--------------------------------
- Installed this build *without giving the app time to lock*
- Opened app > Upgrade dialog
- Selecting `Continue` brings user to signin
- Signed in successfully with OAuth ✅

--------------------------------
- Installed this build *without giving the app time to lock*
- Opened app > Upgrade dialog
- Backgrounded app
- re-opened app, dialog still present
- Selecting `Continue` brings user to signin
- Signed in successfully with OAuth ✅

--------------------------------
- Installed this build *without giving the app time to lock*
- Opened app > Upgrade dialog
- Hard quit app
- re-opened app, dialog re-appeared
- Selecting `Continue` brings user to signin
- Signed in successfully with OAuth ✅

--------------------------------
- Installed this build *without giving the app time to lock*
- Opened app > Upgrade dialog
- Selecting `Continue` brings user to signin
- Hard quit app
- Returning to app, no dialog shown ❓ (product question? this is easier technically, and the user has already selected `Continue` to signify understanding)
- Signed in successfully with OAuth ✅

*happy testing!*

**Comment by devinreams** [comment]
@sashei I re-installed Build 1717 and signed in, then manually locked it. After I upgraded to 1867 and opened the app it was stuck on a black screen. Maybe because I had it locked?

**Comment by sashei** [comment]
@devinreams sure; I was hopeful it had been fixed with the autolock work but maybe not. I will try to repro what @nickbrandt is seeing.

**Comment by sashei** [comment]
@nickbrandt @sandysage and anyone else who is testing this -- can you include the pre-OAuth version number that you install when doing testing? tyvm!

**Comment by nickbrandt** [comment]
@sashei I used build 1664 the first time and tried 1717 the second I believe.

File: lockbox-ios/Presenter/RootPresenter.swift [hunk 1]
```diff
@@ -62,39 +75,32 @@ class RootPresenter {
...
-                        break
-                    }
+        Observable.combineLatest(self.dataStore.locked, self.accountStore.hasOldAccountInformation, self.dataStore.storageState)
+                .filter { !$0.1 && $0.2 != LoginStoreState.Unprepared }
+                .map { $0.0 }
+                .distinctUntilChanged()
+                .subscribe(onNext: { locked in
+                    let route: RouteAction = locked ? LoginRouteAction.welcome : MainRouteAction.list
+
+                    self.routeActionHandler.invoke(route)
```
Comment (jhugman) [inline lockbox-ios/Presenter/RootPresenter.swift]: This looks… brittle.

I'm not sure I understand why we don't separate out the `hasOldAccountInformation` to a `take(1)`.

File: lockbox-ios/Presenter/WelcomePresenter.swift [hunk 2]
```diff
@@ -87,34 +90,24 @@ class WelcomePresenter {
...
+                            AlertActionButtonConfiguration(
+                                    title: Constant.string.continueText,
+                                    tapObserver: self.oauthLoginConfirmationObserver,
+                                    style: .default)
+                        ],
+                        title: Constant.string.reauthenticationRequired,
+                        message: Constant.string.appUpdateDisclaimer,
+                        style: .alert)
+            }
+            .disposed(by: self.disposeBag)
```
Comment (jhugman) [inline lockbox-ios/Presenter/WelcomePresenter.swift]: Rx makes this really ugly. :(

File: lockbox-ios/Store/AccountStore.swift [hunk 3]
```diff
@@ -80,14 +87,28 @@ class AccountStore {
...
 extension AccountStore {
+    private func checkOldAccount() {
+        if self.keychainWrapper.string(forKey: KeychainKey.email.rawValue) != nil ||
+                self.keychainWrapper.string(forKey: KeychainKey.displayName.rawValue) != nil ||
+                self.keychainWrapper.string(forKey: KeychainKey.avatarURL.rawValue) != nil {
+            self._oldAccountPresence.onNext(true)
+        } else {
+            self._oldAccountPresence.onNext(false)
+        }
+    }
```
Comment (jhugman) [inline lockbox-ios/Store/AccountStore.swift]: This feels like it should go in a more central place, and the upgrade should go in a `LifecycleAction.upgrade(old: Int, new: Int)` action.

File: lockbox-ios/View/RootView.swift [hunk 4]
```diff
@@ -36,7 +36,9 @@ class RootView: UIViewController, RootViewProtocol {
 
     init() {
         super.init(nibName: nil, bundle: nil)
-        self.presenter = RootPresenter(view: self)
+        if !isRunningTest {
+            self.presenter = RootPresenter(view: self)
+        }
```
Comment (jhugman) [inline lockbox-ios/View/RootView.swift]: How does this interact with other XCUITests?

**Comment by changecourse** [comment]
- Removed all previous builds, installed latest branch of this PR (build 1915) and was presented with "No Matching Entries" screen after successful authentication. (perhaps also related to #592)  

- Deleted previous app, installed Build 1664 > signed in > entry list populated as expected

- Upgraded to OAuth Branch (build 1915) > Presented with lock screen > Unlocked and Presented with Entry List, no dialog for upgrade, already signed in... I'm confused.

**Comment by nickbrandt** [comment]
@sashei 
After testing this again, I was able to get the prompt as desired upon upgrading. ✅ 
**However**, upon clicking "Continue" from the prompt, it directs me to the FxA sign-in as expected, but if I cancel out of that it brings me to the Lock screen, where I can use biometrics and get back into the app (without re-signing in). ❌

**Comment by sashei** [comment]
@nickbrandt @sandysage @changecourse great, thank you for checking this out, I'll explore those issues.

One nice result of @jhugman 's architecture change (realized this when I was testing locally but forgot to share more widely) is that you can upgrade from *any* branch to this one and (ideally) see the upgrade dialog. So you can install this branch over your install of master and don't have to do the disconnecting song & dance

**Comment by sashei** [comment]
Another cool thing is that you get No Matching Entries if completing the flow properly ¯\_(ツ)_/¯

I'll get to this later this evening or early tomorrow, gotta go do some other things early afternoon today.

**Comment by sashei** [comment]
@nickbrandt @changecourse @sandysage on further thought, the most accurate representation of the upgrade state would be upgrading from what's currently in production. despite this technically working with other builds when you upgrade, I think we'd be better served by doing the full upgrade from latest `production` build (which I believe is 1664). carry on, sorry for the mixed messaging :(

**Comment by sandysage** [comment]
Success for me!

* Downloaded build 1664 > Signed In  ✅
* Left the app running and unlocked > downloaded build 1925 "lockbox" > opened app > app locked with "Reauthentication Required" message ✅
* New OAuth login flow with email address already populated ✅

I did get the "welcome to Firefox Lockbox" screen. Is there any way to distinguish these upgrades as different from new registrations? If so, perhaps we need a follow up story to not display this welcome screen for 'existing' users.

**Comment by changecourse** [comment]
I'm still having issues, though it may be related to FxA... 

__Prior to testing, I removed all previous builds and removed all instances of auth from FxA's Device list

- Downloaded build 1664 > Signed In > Asked to Confirm Email, but upon going to email, not given option to confirm (New Sign In email rather than Confirm Sign-In) Only way to see entries in list was to force quit and reopen app
- Downloaded build 1928 > Splash Screen > Entries List (No dialog, no re-sign in required)

**Comment by nickbrandt** [comment]
Getting some weird behavior when testing this...

Initially tested 3 scenarios as follows, all successful:
- Downloaded 1664 > Signed In
- Tested: backgrounding app / locking app first / force quitting app scenarios
- After updating to this branch, each scenario resulted in being directed to the Welcome screen with the prompt as desired ✅ 
- 'Continue' directs to FxA sign-in ✅ 
- Closing out of FxA sign-in directs back to the Welcome screen, requiring you to sign in ✅ 

I decided to test once more, as I couldn't remember if the user was re-prompted when closing out of FxA sign-in...
This time I was directed to the Welcome screen but without the prompt ❌ 
I then went through each scenario again and each time did not receive the prompt ❌ 
In the scenario where I lock the app first, once I updated to this branch it didn't even navigate me to the Welcome screen, just brought me to unlock screen ❌ 

Thought we were there 😢

**Comment by sashei** [comment]
@sandysage seeing the `Getting Started` screen when the dialog is displayed is part of the designs on this story. there is a way to distinguish between existing-but-upgrading vs. first-time users, but implementing another screen for the former would be follow-up engineering (and presumably design) task.

**Comment by sashei** [comment]
@nickbrandt @changecourse just double checking, did y'all disconnect each time from the signed-in oauth version before installing #1664? I think this step is still required.

**Comment by sashei** [comment]
Also, the user will not be shown the dialog again after they hit `Continue` once, even when returning to the Welcome screen.

**Comment by changecourse** [comment]
Ok. Maybe now it's working as expected? I think. 

- Downloaded 1933 > Signed In > Disconnected
- Downloaded 1664 > Signed In > Stuck in Email Confirmation State
- Downloaded 1933 > Received Prompt > Email Preserved > Signed In > Entries List as Expected

**Comment by nickbrandt** [comment]
@sashei Ah, gotcha.

Re-ran through the same scenarios (disconnecting before each), this time successful every time! ✅  🥂

**Comment by sashei** [comment]
😅😅😅

**Comment by devinreams** [comment]
FWIW this worked for me on the first try on my local iPhone 8 Plus simulator ✅ 

But I did still encounter the "No matching entries" bug when disconnecting and re-connecting a couple times so left that open over here: #592 

Nice work and thank you very much, @sashei 🙇

**Pull request merged**